rados/read_op.go

```diff
@@ -0,0 +1,84 @@
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
+//
+import "C"
+
+import (
+	"unsafe"
+)
+
+// ReadOp manages a set of discrete object read actions that will be performed
+// together atomically.
+type ReadOp struct {
+	operation
+	op C.rados_read_op_t
+}
+
+// CreateReadOp returns a newly constructed read operation.
+func CreateReadOp() *ReadOp {
+	return &ReadOp{
+		op: C.rados_create_read_op(),
+	}
+}
+
+// Release the resources associated with this read operation.
+func (r *ReadOp) Release() {
+	C.rados_release_read_op(r.op)
+	r.op = nil
+	r.free()
+}
+
+// Operate will perform the operation(s).
+func (r *ReadOp) Operate(ioctx *IOContext, oid string, flags OperationFlags) error {
+	if err := ioctx.validate(); err != nil {
+		return err
+	}
+
+	cOid := C.CString(oid)
+	defer C.free(unsafe.Pointer(cOid))
+
+	ret := C.rados_read_op_operate(r.op, ioctx.ioctx, cOid, C.int(flags))
+	return r.update(readOp, ret)
+}
+
+func (r *ReadOp) operateCompat(ioctx *IOContext, oid string) error {
+	switch err := r.Operate(ioctx, oid, OperationNoFlag).(type) {
+	case nil:
+		return nil
+	case OperationError:
+		return err.OpError
+	default:
+		return err
+	}
```

Comment:
I don't think the errors.As construct helps much here and it pulls in a module dependency. Plus I doubt we'll be adding lots of new error types to this so the As check also feels a bit of a YAGNI.

Reply:
I'm confused. Which module are you talking about? We use `errors` all over the place. So it looks better, will not break if we might wrap errors in the future, and we are avoiding issues with linters (as just happened with CSI), and I really don't see any downsides. 🤷‍♂️

Reply:
I just really don't like relying on "new" language features unless its absolutely needed. I'm just very much in the habit of needing to support older language versions for other projects. Since its a bit of a wash which is clearer/simpler I figured why add the extra dependency (on the call - not just the package). I also don't think we'll be wrapping it any time soon. Do you mind if you solicit the other reviewers to get more opinions before deciding on this one?

Reply:
I think the usage of the switch on error is fine here. Errors come from the same package, so I do not see the strong dependency as a problem. My opinion would be different if the error comes from an other package.